fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java

```diff
@@ -0,0 +1,128 @@
+ */
+
+package com.apple.foundationdb.record.query.plan.plans.visitor;
+
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.RecordType;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.query.plan.IndexKeyValueToPartialRecord;
+import com.apple.foundationdb.record.query.plan.RecordQueryPlanner;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryCoveringIndexPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryFetchFromPartialRecordPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlanWithIndex;
+import com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Visitor interface for performing substitution-type rules on {@link RecordQueryPlan}s.
+ */
+public abstract class RecordQueryPlannerSubstitutionVisitor {
+    @Nonnull
+    private static final Logger logger = LoggerFactory.getLogger(RecordQueryPlanner.class);
+
+    @Nonnull
+    protected final RecordMetaData recordMetadata;
+    @Nullable
+    private final KeyExpression commonPrimaryKey;
+
+    public RecordQueryPlannerSubstitutionVisitor(@Nonnull RecordMetaData recordMetadata, @Nullable KeyExpression commonPrimaryKey) {
+        this.recordMetadata = recordMetadata;
+        this.commonPrimaryKey = commonPrimaryKey;
+    }
+
+    public static RecordQueryPlan applyVisitors(@Nonnull RecordQueryPlan plan, @Nonnull RecordMetaData recordMetaData, @Nullable KeyExpression commonPrimaryKey) {
+        return plan.accept(new UnorderedPrimaryKeyDistinctVisitor(recordMetaData, commonPrimaryKey))
+                .accept(new UnionVisitor(recordMetaData, commonPrimaryKey))
+                .accept(new IntersectionVisitor(recordMetaData, commonPrimaryKey))
+                .accept(new UnorderedPrimaryKeyDistinctVisitor(recordMetaData, commonPrimaryKey));
+    }
+
+    @Nonnull
+    public abstract RecordQueryPlan visit(@Nonnull RecordQueryPlan recordQueryPlan);
+
+    @Nullable
+    protected RecordQueryPlan removeIndexFetch(@Nonnull RecordQueryPlan plan) {
+        if (plan instanceof RecordQueryPlanWithIndex) {
+            RecordQueryPlanWithIndex indexPlan = (RecordQueryPlanWithIndex) plan;
+            Index index = recordMetadata.getIndex(indexPlan.getIndexName());
+            final Collection<RecordType> recordTypes = recordMetadata.recordTypesForIndex(index);
+            final IndexKeyValueToPartialRecord keyValueToPartialRecord = generateIndexKeyValueToPartialRecord(indexPlan);
+            if (keyValueToPartialRecord != null && recordTypes.size() == 1) {
```

Comment:
There is duplication between here and `generateIndexKeyValueToPartialRecord`:
* get index from metadata
* get record types from index
* assert only one record type

Since the method is private, it seems like a little bit of refactoring could avoid that.

Reply:
Extracted into a private method.

Reply:
Did this get pushed?
In addition to the source code duplication, can't we avoid doing these more than one in quick succession?

Reply:
No, not yet. I'm still getting the comparison key validation logic working. I think I've found a way to pull this out into a single spot per evaluation.